pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,98 @@
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	affinity, err := unmarshalAffinity(settings.ClusterAgentDefaultAffinity.Get())
+	if err != nil {
+		return settings.GetClusterAgentDefaultAffinity()
+	}
+
+	return affinity
+}
+
+// GetClusterAgentResourceRequirements returns resource requirements (cpu, memory) for the cluster agent if it has been
+// user defined. If not, nil is returned.
+func GetClusterAgentResourceRequirements(cluster *v3.Cluster) *corev1.ResourceRequirements {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements
+	}
+
+	return nil
+}
+
+// GetFleetAgentTolerations returns additional tolerations for the fleet agent if it has been user defined. If not,
+// then nil is returned.
+func GetFleetAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetFleetAgentAffinity returns node affinity for the fleet agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetFleetAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	affinity, err := unmarshalAffinity(settings.FleetAgentDefaultAffinity.Get())
+	if err != nil {
+		return settings.GetFleetAgentDefaultAffinity()
+	}
+
+	return affinity
+}
+
+// GetFleetAgentResourceRequirements returns resource requirements (cpu, memory) for the fleet agent if it has been
+// user defined. If not, nil is returned.
+func GetFleetAgentResourceRequirements(cluster *v3.Cluster) *corev1.ResourceRequirements {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.OverrideResourceRequirements != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.OverrideResourceRequirements
+	}
+
+	return nil
+}
+
+// ummarshalAffinity returns an unmarshalled v1 node affinity object. If unable to be unmarshalled, it returns nil
+// and an error.
+func unmarshalAffinity(affinity string) (*corev1.Affinity, error) {
+	var affinityObj corev1.Affinity
+	err := json.Unmarshal([]byte(affinity), &affinityObj)
+	if err != nil {
+		logrus.Errorf("failed to unmarshal node affinity: %v", err)
```

Comment:
Is the error logging here required? We should be bubbling these errors back upwards to the calling function to let it wrap the error/provide context.